### PR TITLE
Preparation of bring overrides authoring prototype to development

### PR DESCRIPTION
**Pull request by chiyenteng**
Modify InstanceUpdateExecutor, PrefabDomUtils, and PrefabFocusPublicInterface for bring overrides authoring prototype to development later.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Entity/EditorEntityModel.cpp [hunk 1]
```diff
@@ -362,6 +362,7 @@ namespace AzToolsFramework
             if (parentInfo.HasChild(childId))
             {
                 ReparentChild(childId, entityId, AZ::EntityId());
+                AzToolsFramework::ToolsApplicationRequests::Bus::Broadcast(&ToolsApplicationRequests::RemoveDirtyEntity, childId);
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Entity/EditorEntityModel.cpp:365]: Aren't these changes to `EditorEntityModel` in development already?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp [hunk 2]
```diff
@@ -179,42 +305,9 @@ namespace AzToolsFramework
...
-                                "InstanceUpdateExecutor::UpdateTemplateInstancesInQueue - "
-                                "Could not load Instance DOM from the top level ancestor's DOM.");
-
-                            isUpdateSuccessful = false;
-                            continue;
-                        }
-
-                        PrefabDomValueConstReference instanceDomFromRoot = *instanceDomFromRootValue;
-                        if (!instanceDomFromRoot.has_value())
+                        // TODO - Add relevant comment.
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp:308]: nit: either remove the comment, or replace it with something useful :)

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabFocusPublicInterface.h [hunk 3]
```diff
@@ -56,6 +66,17 @@ namespace AzToolsFramework::Prefab
...
+        //! Returns the current focus mode.
+        virtual PrefabEditScope GetPrefabEditScope(AzFramework::EntityContextId entityContextId) const = 0;
+
+        //! Sets the current focus mode.
+        virtual void SetPrefabEditScope(AzFramework::EntityContextId entityContextId, PrefabEditScope mode) = 0;
+
+        // TEMP
+        virtual int GetOpenInstanceMode() = 0;
+        virtual bool GetAllowContextMenuInstanceExpanding() = 0;
+        virtual bool GetContainerStepByStepSelection() = 0;
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabFocusPublicInterface.h:79]: Could you add short doc comments here? Thanks :D

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp [hunk 4]
```diff
@@ -94,9 +121,108 @@ namespace AzToolsFramework
...
+                }
+            }
+            // If our instance is the focused instance, fix the container
+            else if (&focusedInstance->get() == instance)
             {
-                return entry == instance;
-            });
+                // Replace the container entity with the one as seen by the root
+                // TODO - this function should only replace the transform!
+                ReplaceFocusedContainerTransformAccordingToRoot(instance, instanceDom);
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp:216]: Is this change to ReplaceFocusedContainerTransformAccordingToRoot planned as a task?

**Review by AMZN-daimini — COMMENTED** [review]
I don't feel confident approving this since I wrote the original code, but I added a few comments :)

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabDomUtils.cpp [hunk 5]
```diff
@@ -451,6 +451,45 @@ namespace AzToolsFramework
...
+                rapidjson::StringBuffer prefabBuffer;
+                rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(prefabBuffer);
+                prefabDomValue.Accept(writer);
+
+                return prefabBuffer.GetString();
+            }
+
+            
+            const Instance* ClimbUpToTargetInstance(
+                const Instance* startInstance, const Instance* targetInstance, AZStd::string& aliasPath)
```
[the remark this answers is not in the document]
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabDomUtils.cpp:466]: You mean `RelativePathToTargetInstance`?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp [hunk 6]
```diff
@@ -225,15 +318,12 @@ namespace AzToolsFramework
                             continue;
                         }
 
-                        // If a link was created for a nested instance before the changes were propagated,
-                        // then we associate it correctly here
-                        instanceDomFromRootDocument.CopyFrom(instanceDomFromRoot->get(), instanceDomFromRootDocument.GetAllocator());
-                        if (PrefabDomUtils::LoadInstanceFromPrefabDom(*instanceToUpdate, newEntities, instanceDomFromRootDocument, PrefabDomUtils::LoadFlags::UseSelectiveDeserialization))
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp:231]: Need to add back the UseSelectiveDeserialization flag here.

**Review by srikappa-amzn — CHANGES_REQUESTED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp [hunk 7]
```diff
@@ -94,9 +121,108 @@ namespace AzToolsFramework
...
+
+                // Paste the focused instance container dom as seen from the root into instanceDom.
+                AZStd::string containerName =
+                    AZStd::string::format("/%s", PrefabDomUtils::ContainerEntityName);
+                PrefabDomPath containerPath(containerName.c_str());
+                containerPath.Set(focusedInstanceDom, containerDom, focusedInstanceDom.GetAllocator());
+            }
+        }
+
+        bool InstanceUpdateExecutor::GenerateInstanceDomAccordingToCurrentFocus(const Instance* instance, PrefabDom& instanceDom)
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutor.cpp:160]: This feels like the InstanceUpdateExecutor is trying to too much. Classes should follow single-responsibility-principle to keep systems maintainable. The responsibility of this class should be, "given a queue of instances, I'll propagate given Doms to them". It doesn't even need to know what a focus mode is. So I recommend you create another class that outputs a Dom to be used for propagation by taking an instance as input. The class (may be InstanceDomGenerator) can do focus specific logic and fetch the required dom.

**Review by srikappa-amzn — CHANGES_REQUESTED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/Entity/EditorEntityModel.cpp (hunk 1 above)
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Entity/EditorEntityModel.cpp:365]: I don't know why but when I tried to break the original commit down this old change just got mixed into the PR.

[the remark this answers is not in the document]
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Entity/EditorEntityModel.cpp:365]: I don't know why but when I tried to break the original commit down this old change just got mixed into the PR.

**Pull request closed without merging**